core/market/src/negotiation/provider.rs

```diff
@@ -228,37 +230,110 @@ impl ProviderBroker {
+        // TODO: Reverse state to `Pending` in case of error (reverse under lock).
+        // TODO: During reversing it can turn out, that we are in `Cancelled` or `Approved` state
+        //  since we weren't under lock during `self.api.approve_agreement` execution. In such a case,
+        //  we shouldn't return error from here.
 
-        counter!("market.agreements.provider.approved", 1);
         log::info!(
-            "Provider {} approved Agreement [{}].",
+            "Provider {} approved Agreement [{}]. Waiting for commit from Requestor [{}].",
             id.display(),
-            &agreement_id,
+            &agreement.id,
+            agreement.requestor_id
         );
-        if let Some(session) = app_session_id {
-            log::info!(
-                "AppSession id [{}] set for Agreement [{}].",
-                &session,
-                &agreement_id
-            );
-        }
+
+        // TODO: Here we must wait until `AgreementCommitted` message, since `approve_agreement`
+        //  is supposed to return after approval.
+        // TODO: Waiting should set timeout.
+        // TODO: What in case of timeout?? Reverse to `Pending` state?
+        // Note: This function isn't responsible for changing Agreement state to `Approved`.
+
+        // TODO: Check Agreement state here, because it could be `Cancelled`.
         return Ok(ApprovalResult::Approved);
     }
 }
 
+async fn on_agreement_committed(
+    broker: CommonBroker,
+    caller: String,
+    msg: AgreementCommitted,
+) -> Result<(), CommitAgreementError> {
+    let agreement_id = msg.agreement_id.clone();
+    let caller_id = CommonBroker::parse_caller(&caller)?;
+    agreement_committed(broker, caller_id, msg)
+        .await
+        .map_err(|e| CommitAgreementError::Remote(e, agreement_id))
+}
+
+async fn agreement_committed(
+    broker: CommonBroker,
+    caller: NodeId,
+    msg: AgreementCommitted,
+) -> Result<(), RemoteCommitAgreementError> {
+    let dao = broker.db.as_dao::<AgreementDao>();
+    let agreement = {
+        let _hold = broker.agreement_lock.lock(&msg.agreement_id).await;
+
+        // Note: we still validate caller here, because we can't be sure, that we were caller
+        // by the same Requestor.
+        let agreement = match dao
+            .select(&msg.agreement_id, Some(caller), Utc::now().naive_utc())
+            .await
+            .map_err(|e| AgreementError::Get(msg.agreement_id.to_string(), e))?
+        {
+            None => Err(AgreementError::NotFound(msg.agreement_id.to_string()))?,
+            Some(agreement) => agreement,
+        };
+
+        // Note: We can find out here, that our Agreement is already in `Cancelled` state, because
+        // Requestor is allowed to call `cancel_agreement` at any time, before we commit Agreement.
+        // In this case we should return here, but we still must call `notify_agreement` to wake other threads.
```

Comment:
I think they have been already wakened up as a side effect of `cancel_agreement` or other `commit_agreement` which is also possible in between. But `notifying` more is better than less.